tests/chainer_tests/test_variable.py

```diff
@@ -983,358 +977,314 @@ def v():
-            x_var.grad_var = chainer.Variable(gx, requires_grad=requires_grad)
+        x, gx = src_backend_config.get_array((x, gx))
 
-        x_var.to_cpu()
+        # *_src arrays are kept for the post transfer checks.
+        self.x_src = x
+        self.gx_src = gx
 
-        assert x_var.xp is np
-        assert x_var._has_chainerx_array is False
-        assert x_var.node is not None
-        assert isinstance(x_var.data, np.ndarray)
-        assert x.shape == x_var.shape
-        assert x.dtype == x_var.dtype
-        np.testing.assert_array_equal(
-            backend.CpuDevice().send(x_var.data), backend.CpuDevice().send(x))
-
-        if set_grad_var:
-            assert isinstance(x_var.grad, np.ndarray)
-            assert gx.shape == x_var.grad.shape
-            assert gx.dtype == x_var.grad.dtype
-            np.testing.assert_array_equal(
-                backend.CpuDevice().send(x_var.grad),
-                backend.CpuDevice().send(gx))
-            assert x_var.grad_var is not None
-            assert x_var.grad_var.node is not None
-        else:
-            assert x_var.grad is None
-            assert x_var.grad_var is None
-
-        orig_xp = backend.get_array_module(x, gx)
-        if orig_xp is np:
-            assert x_var.data is x
-            assert x_var.grad is gx
-        else:
-            assert x_var.data is not x
-            assert not set_grad_var or x_var.grad is not gx
-
-        assert x_var.xp is not chainerx
-        assert x_var._has_chainerx_array is False
+        x = chainer.Variable(x)
+        gx = chainer.Variable(gx)
 
-    def test_to_cpu_from_cpu(self):
-        self.check_to_cpu(self.x, self.gx)
+        if self.x_with_graph:
+            x = 2 * x
 
-    @attr.gpu
-    def test_to_cpu_from_gpu(self):
-        self.check_to_cpu(cuda.to_gpu(self.x), cuda.to_gpu(self.gx))
+        if self.x_with_grad:
+            x.grad_var = gx
 
-    @attr.chainerx
-    def test_to_cpu_from_chainerx(self):
-        self.check_to_cpu(
-            chainerx.array(self.x),
-            chainerx.array(self.gx),
-            requires_grad=False)
+        self.x = x
+        self.gx = gx
 
-    @attr.chainerx
-    def test_to_cpu_from_chainerx_requiring_grad(self):
-        with self.assertRaises(RuntimeError):
-            self.check_to_cpu(
-                chainerx.array(self.x),
-                chainerx.array(self.gx),
-                requires_grad=True)
+    def to_device_meth(self, dst_device, allow_unchain):
+        # Call `self.x.to_device`, `self.x.from_chx`, etc.
+        raise NotImplementedError('Must be overridden.')
 
+    def check_to_device_meth_devices(self, src_device, dst_device):
+        # Called after the device transfer, check `self.x`, etc.
+        raise NotImplementedError('Must be overridden.')
 
-@testing.parameterize(
-    {'x_shape': (10,)},
-    {'x_shape': ()},
-)
-@attr.gpu
-class TestVariableToGpu(unittest.TestCase):
+    def check_to_device_meth_ndarrays(self):
```

Comment:
Please write a comment to this method too.
The distinction from `check_to_device_meth_devices` is not clear.

Reply:
Commented. What do you think?

Reply:
Looks good!